codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt

```diff
@@ -141,12 +138,15 @@ class ServerProtocolTestGenerator(
                 // These tests are broken because they are missing a target header.
                 FailingTest.RequestTest(AWS_JSON_10, "AwsJson10ServerPopulatesNestedDefaultsWhenMissingInRequestBody"),
                 FailingTest.RequestTest(AWS_JSON_10, "AwsJson10ServerPopulatesDefaultsWhenMissingInRequestBody"),
+                FailingTest.RequestTest(REST_JSON, "RestJsonServerPopulatesDefaultsWhenMissingInRequestBody"),
```

Comment:
`RestJsonServerPopulatesDefaultsWhenMissingInRequestBody` fails because `defaultNullDocument` has a `null` default, so the `hasNonNullDefault` check here:

https://github.com/smithy-lang/smithy-rs/blob/9af72f5f2a04e941dd7f0f17a292e4396efdfda6/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGenerator.kt#L537

is `false` and so [`aws_smithy_types::Document::Null`](https://docs.rs/aws-smithy-types/latest/aws_smithy_types/enum.Document.html#variant.Null) is not used by the instantiator.

But even if we improved the check and special-cased `document` member shapes, there's an additional problem: the same check is used by Smithy itself [here](https://github.com/smithy-lang/smithy/blob/efcb39cf19f4cc71be65b64d2337802d628b8571/smithy-model/src/main/java/software/amazon/smithy/model/knowledge/NullableIndex.java#L142) when determining whether the structure member should be optional or not. Since the node for the `@default` value of the member shape is `null`, Smithy says the structure member should be optional. But the generated code for default value setting expects the member to be non-`Option`al, since there's a default value.

In general, this begs the question of whether a required (non-`Option`al) `document` shape makes sense at all. The user can always set it to [`aws_smithy_types::Document::Null`](https://docs.rs/aws-smithy-types/latest/aws_smithy_types/enum.Document.html#variant.Null) and convey the same semantics as if shape were nullable and `None` had been set for it. To belabor the point, consider this model:

```
structure OperationInput {
    @required
    document: Document
}
```

This gets generated in Rust as:

```
struct OperationInput {
    document: aws_smithy_types::Document
}
```

The user can instantiate this as:

```
let op_input = OperationInput::builder()
    .document(aws_smithy_types::Document::Null)
    .build();
```

Which will get serialized on the wire as `{ document: null }` or `{ }`, depending on whether the protocol serializes nulls or not, respectively. Either way, we're allowing the user to not provide a value for a required member shape, violating the model.

Reply:
Thanks for the detailed explanation! Moved the test to the bottom of the list and added a link to this discussion for more context.